velox/connectors/hive/iceberg/IcebergPartitionIdGenerator.h

```diff
@@ -0,0 +1,132 @@
+  /// (identity, bucket, truncate, year, etc.), and optional parameters.
+  /// Multiple transforms on the same column are allowed with restrictions
+  /// enforced by IcebergPartitionSpec validation.
+  /// See IcebergPartitionSpec::checkCompatibility for detail.
+  /// @param maxPartitions Maximum number of distinct partitions allowed.
+  /// Exceeding this limit will cause a VeloxUserError.
+  /// @param connectorQueryCtx Connector query context for expression evaluation
+  /// and memory allocation.
+  IcebergPartitionIdGenerator(
+      const RowTypePtr& inputType,
+      std::vector<column_index_t> partitionChannels,
+      IcebergPartitionSpecPtr partitionSpec,
+      uint32_t maxPartitions,
+      const ConnectorQueryCtx* connectorQueryCtx);
+
+  /// Generate sequential partition IDs for all rows in the input vector.
+  ///
+  /// For each row, applies the partition transforms, hashes the transformed
+  /// values, and maps to a sequential partition ID. New partitions are created
+  /// as needed (up to maxPartitions limit). Partition values are stored for
+  /// later partition name generation.
+  ///
+  /// @param input Input RowVector containing all columns. Must have columns
+  /// at indices specified by partitionChannels.
+  /// @param result Output vector of partition IDs, one per input row.
+  /// Resized to match input size. Values are 0-based sequential IDs.
+  void run(const RowVectorPtr& input, raw_vector<uint64_t>& result) override;
+
+  /// Generate the partition name for the given partition ID.
+  ///
+  /// Constructs an Iceberg-compliant partition path string in the format:
+  /// "key1=value1/key2=value2/..." where keys are partition column names for
+  /// identity transform or column_transform for non-identity transforms and
+  /// values are the human-readable string representations of the transformed
+  /// partition values. Special characters are URL-encoded per
+  /// java.net.URLEncoder.encode().
+  ///
+  /// In typical usage (e.g., HiveDataSink::appendWriter), this is called
+  /// once per partition ID when creating a new writer for that partition.
+  /// It performs string formatting and URL encoding on each call, so callers
+  /// should cache the result if the same partition name is needed
+  /// multiple times.
+  ///
+  /// @param partitionId Sequential partition ID (0-based) returned by run().
+  /// Must be less than the number of partitions created so far.
+  /// @return Partition path string suitable for use in file paths.
+  std::string partitionName(uint32_t partitionId) const override;
+
+  /// Return the RowVector containing transformed partition values.
+  /// Each row in this vector corresponds to a partition ID (row index =
+  /// partition ID). Each column contains the transformed values for that
+  /// partition column.
+  ///
+  /// @return RowVector with one column per transformed column, columns in same
+  /// order as IcebergPartitionSpec::fields. Returns nullptr if no partitions
```

Comment:
> Returns nullptr if no partitions

Is this accurate? The implementation suggests that return value is never nullptr.